### PR TITLE
IBM Granite Architecture

### PR DESCRIPTION
**Pull request by gabe-l-hart**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

## Description

This PR introduces the `granite` model architecture from IBM. It emulates the `transformers` changes in [this PR](https://github.com/huggingface/transformers/pull/31502).

The `granite` architecture is closely related to `llama` and therefore reuses many of the existing `llama` code paths. The primary differences are:

1. `embeddings_multiplier`: A scaling factor applied to the input embeddings
2. `attention_multiplier`: A configurable scaling factor applied to the output of each attention block that replaces `1 / sqrt(n_embd_head)`
3. `residual_multiplier`: A scaling factor applied when recombining the attention output with the residual and again when combining the FFN output with the residual
4. `logits_scale`: A scaling factor applied to the final logits before decoding

## Design Decisions/Questions

Since this is my first model addition, there were a couple of design choices that I took my best guess on, but would like to point out for further discussion:

* In `convert_hf_to_gguf.py`, I derived `GraniteModel` from `LlamaModel` to avoid copy-pasting the defaults there. I see that other models derive directly from `Model` and avoid inheritance, so I would be open to following that pattern if it's preferred.
* Parameter naming (`llama.cpp`):
    * There was no clear existing parameter for `embeddings_multiplier`/`attention_multiplier`/`residual_multiplier`, so I opted to use the same names that are used in `transformers`. This does, however, leave an ugly asymmetry with `logits_scale`, so I thought about `*_scale` as well.
    * I noticed that in `llama_hparams`, some variables have a prefix character, while others don't. I opted to follow the `f_` prefix notation, but am happy to remove that if it's not to convention.
* `build_llama`: Similar to using `LlamaModel` as a base class, I opted to make small tweaks to the `build_llama` function rather than copy-paste into a clean-room `build_granite` function with the small deltas.
* Unit Testing: I see that there are not unit tests for each architecture and other model addition PRs don't seem to include them, so I did not either

## Testing

To test the model architecture, I am working with IBM's preview model [ibm/PowerLM-3b](https://huggingface.co/ibm/PowerLM-3b).

```sh
huggingface-cli download ibm/PowerLM-3b --local-dir $HOME/models/powerlm-3b
```

### Conversion

```sh
python convert_hf_to_gguf.py --verbose $HOME/models/powerlm-3b/
```

### Simple Execution

```sh
./build/bin/llama-simple -m $HOME/models/powerlm-3b/powerlm-3B-F16.gguf -p "Write a code to find the maximum value in a list of numbers." -n 100
```

> ASSISTANT: Here's a code to find the maximum value in a list of numbers:
> ```
> def find_max(numbers):
>  max_value = numbers[0]
>  for num in numbers:
>  if num > max_value:
>  max_value = num
>  return max_value
> ```
> This code defines a function called `find_max` that takes a list of numbers as

### Quantization

```sh
./build/bin/llama-quantize $HOME/models/powerlm-3b/powerlm-3B-F16.gguf Q4_K_M

# Simple with the quantized model
./build/bin/llama-simple -m /Users/ghart/models/powerlm-3b/ggml-model-Q4_K_M.gguf -p "Write a code to find the maximum value in a list of numbers." -n 100
```

> ASSISTANT: Sure, here's a code to find the maximum value in a list of numbers:
> ```
> def find_max(numbers):
>  max_value = numbers[0]
>  for num in numbers:
>  if num > max_value:
>  max_value = num
>  return max_value
> ```
> This code defines a function called `find_max` that takes a list of

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Nice PR - good job!

> In convert_hf_to_gguf.py, I derived GraniteModel from LlamaModel to avoid copy-pasting the defaults there. I see that other models derive directly from Model and avoid inheritance, so I would be open to following that pattern if it's preferred.

I believe this is OK. Ping @compilade to confirm.

> There was no clear existing parameter for embeddings_multiplier/attention_multiplier/residual_multiplier, so I opted to use the same names that are used in transformers. This does, however, leave an ugly asymmetry with logits_scale, so I thought about *_scale as well.

Yes, it's a tough call. Just from `llama.cpp` PoV, it's best to have all factors with the same suffix/prefix for consistency. But on the other hand, matching the names with existing frameworks like `transformers` is also valuable. We haven't been consistent either way, so I guess we can leave it as proposed.

> I noticed that in llama_hparams, some variables have a prefix character, while others don't. I opted to follow the f_ prefix notation, but am happy to remove that if it's not to convention.

Again, we have not been very consistent in this regard. Adding the `f_` prefix for now is OK and in the future we will try to normalize the names.

> build_llama: Similar to using LlamaModel as a base class, I opted to make small tweaks to the build_llama function rather than copy-paste into a clean-room build_granite function with the small deltas.

Good call.

> Unit Testing: I see that there are not unit tests for each architecture and other model addition PRs don't seem to include them, so I did not either

We don't have unit tests yet for the build code.

File: src/llama.cpp [hunk 1]
```diff
@@ -5368,13 +5401,20 @@ static void llm_load_hparams(
                         // granite uses a vocab with len 49152
                         case 32: model.type = hparams.n_vocab == 49152 ? e_model::MODEL_3B : (hparams.n_vocab < 40000 ? e_model::MODEL_7B : e_model::MODEL_8B); break;
                         case 36: model.type = e_model::MODEL_8B; break; // granite
-                        case 40: model.type = e_model::MODEL_13B; break;
+                        case 40: model.type = (hparams.n_vocab == 49152 || hparams.n_vocab == 49156) ? e_model::MODEL_3B : e_model::MODEL_13B; break;
```
Comment (compilade) [inline src/llama.cpp:5404]: Here, I think it might be more appropriate to use `hparams.n_embd` to differentiate the sizes. `hparams.n_vocab` was used to distinguish between 7B and 8B Llama models because that was the only difference, but between a 3B and a 13B model with the same layer count, I'm sure `n_embd` isn't the same.

```suggestion
                        case 40: model.type = (hparams.n_embd == 2304) ? e_model::MODEL_3B : e_model::MODEL_13B; break;
```

Although maybe this should be a nested switch-case instead.

**Review by compilade — APPROVED** [review]
> In convert_hf_to_gguf.py, I derived GraniteModel from LlamaModel to avoid copy-pasting the defaults there. I see that other models derive directly from Model and avoid inheritance, so I would be open to following that pattern if it's preferred.

This is okay; inheritance where it makes sense is useful and is preferred in some cases (e.g. Bert-like models also make use of it).

**Comment by gabe-l-hart** [comment]
Thanks for the reviews @ggerganov and @compilade! All of your comments make sense. I'll plan to address them first thing on Monday

File: convert_hf_to_gguf.py [hunk 2]
```diff
@@ -3931,6 +3931,34 @@ def prepare_tensors(self):
...
+
+        - No head_dim support
+        - New multiplier params:
+            - attention_multiplier
+            - embedding_multiplier
+            - residual_multiplier
+        - logits_scaling
+        """
+        if head_dim := self.hparams.pop("head_dim", None):
+            logger.warning("Ignoring head_dim (%s) from config for Granite", head_dim)
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply convert_hf_to_gguf.py:4099]: Good question. I'm not intimately familiar with why it was done this way, but in [the `transformers` implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/granite/modeling_granite.py#L279), there is a `ValueError` raised if `head_dim` is not directly computed from `num_heads` and `hidden_size` and it is not supported as a parameter to the [model's config](https://github.com/huggingface/transformers/blob/main/src/transformers/models/granite/configuration_granite.py), so I put this here as a guard against an invalid `config.json` that had it present.

File: src/llama.cpp (hunk 1 above)
Comment (gabe-l-hart) [reply src/llama.cpp:5458]: I took the other suggestion to split this whole switch clause out, so this got quite a bit simpler. Thanks!

File: src/llama.cpp [hunk 3]
```diff
@@ -5367,13 +5400,20 @@ static void llm_load_hparams(
...
                         default: model.type = e_model::MODEL_UNKNOWN;
                     }
                 }
+                // Extra multipliers for Granite architecture
+                if (model.arch == LLM_ARCH_GRANITE) {
+                    ml.get_key(LLM_KV_LOGIT_SCALE, hparams.f_logit_scale);
+                    ml.get_key(LLM_KV_RESIDUAL_MULTIPLIER, hparams.f_residual_multiplier);
+                    ml.get_key(LLM_KV_EMBEDDING_MULTIPLIER, hparams.f_embedding_multiplier);
+                    ml.get_key(LLM_KV_ATTENTION_MULTIPLIER, hparams.f_attention_multiplier);
+                }
```
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply src/llama.cpp]: Good call, this was pretty messy

**Comment by gabe-l-hart** [comment]
Thanks again for the quick reviews! I've addressed them with the following updates:

* Switch from `*_multiplier` to `*_scale` and convert from the `transformers` during model conversion. I like keeping this consistent with the other `hparams` in `llama.cpp`.
* Split out the `LLM_ARCH_GRANITE` clause in `llm_load_hparams` so avoid the added conditionals in the `LLM_ARCH_LLAMA` clause.
